mesonbuild/dependencies/python.py

```diff
@@ -282,7 +282,7 @@ def find_libpy_windows(self, env: 'Environment') -> None:
         arch = detect_cpu_family(env.coredata.compilers.host)
         if arch == 'x86':
             arch = '32'
-        elif arch == 'x86_64':
+        elif arch == 'x86_64' or arch == 'aarch64':
```

Comment:
I think this needs to be refactored a bit. That code checks that we don't build for x86_64 with an x86 libpython, or vice versa, assuming Windows only support x86 arches. But now we should also check we don't mix x86_64 and aarch64, we can't assume that because they are both 64bits it's going to be fine.

I think get_windows_python_arch() should return the full cpu family.

Reply:
Do you mean something like this instead?
```diff
diff --git a/mesonbuild/dependencies/python.py b/mesonbuild/dependencies/python.py
index 478df6ac4..d5484e4a3 100644
--- a/mesonbuild/dependencies/python.py
+++ b/mesonbuild/dependencies/python.py
@@ -191,7 +191,7 @@ class PythonSystemDependency(SystemDependency, _PythonDependencyBase):

         # https://sourceforge.net/p/mingw-w64/mailman/message/30504611/
         # https://github.com/python/cpython/pull/100137
-        if mesonlib.is_windows() and self.get_windows_python_arch() == '64' and mesonlib.version_compare(self.version, '<3.12'):
+        if mesonlib.is_windows() and self.get_windows_python_arch.endswith('64') and mesonlib.version_compare(self.version, '<3.12'):
             self.compile_args += ['-DMS_WIN64']

         if not self.clib_compiler.has_header('Python.h', '', environment, extra_args=self.compile_args):
@@ -232,6 +232,8 @@ class PythonSystemDependency(SystemDependency, _PythonDependencyBase):
             return '32'
         elif self.platform in {'win64', 'win-amd64'}:
             return '64'
+        elif self.platform == 'win-arm64':
+            return 'arm64'
         mlog.log(f'Unknown Windows Python platform {self.platform!r}')
         return None

@@ -284,6 +286,8 @@ class PythonSystemDependency(SystemDependency, _PythonDependencyBase):
             arch = '32'
         elif arch == 'x86_64':
             arch = '64'
+        elif arch == 'aarch64':
+            arch = 'arm64'
         else:
             # We can't cross-compile Python 3 dependencies on Windows yet
             mlog.log(f'Unknown architecture {arch!r} for',
```

Reply:
Return one of x86, x86_64 or aarch64 string, then you can remove all if blocks here and do directly if arch!=pyarch

Reply:
Why wasn't that done in the first place? I don't know any meson internals and don't really want to get into much refactoring or change undocumented return values of public methods...

Reply:
Meson doesn't have public methods because it's a CLI tool. All code is a valid candidate for internal refactoring, if that is beneficial another change you are making and all call sites are updated to keep working.

In the case of dependency classes, these particular instance methods are only allowed to be used within this one class. :D

(I think I dropped the underscore prefix when I moved and heavily refactored the python factory, sorry for the confusion. They don't lack an underscore because they are public. They lack an underscore because the entire project is pseudo-private and the people who frequently work on it "know" that it only makes sense to directly operate on the base classes for SystemDependency or PkgConfigDependency etc.)

...

Why was it originally implemented this way? Good question. It's based on the work in #2697, which was "get the python dependency test to work in MSYS2 CI so this unrelated PR can have clean CI and be merged".

Reply:
OK. Thanks for the explanation.